### PR TITLE
Add test for sort_name, display_name consistency

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Adds a test for an api3 vs 4 anomaly - test expected to fail - fix to be discussed

Before
----------------------------------------
No test

After
----------------------------------------
Test

Technical Details
----------------------------------------
@colemanw what this demonstrates is that if you create an individual knowing just the email in v3 the contact create will take
```
civicrm_api3('Contact', 'create', ['contact_type' => 'Individual', 'email' => 'coleman@example.org']);

```

And populate sort_name & display_name with those. For apiv4 it doesn't which I think needs some form of redress - since I don't think any currently recommended apiv4 pattern does.

The code doing it for apiv3 is in 2 places (both nasty)
1) in the api- - format for the other place https://github.com/civicrm/civicrm-core/blob/c3cea83f4e1f624d698d70a6ee943a655ee3be5f/api/v3/Contact.php#L58
2) in the BAO use the secret code formatting to do magic https://github.com/civicrm/civicrm-core/blob/dcb148082af818050ec0d9c56b04acd37967b34c/CRM/Contact/BAO/Individual.php#L213

Comments
----------------------------------------

### CODE REVIEW

File: tests/phpunit/api/v3/ContactTest.php [hunk 1]
```diff
@@ -1573,6 +1576,38 @@ public function testCreateIndividualWithNameEmail() {
...
+   * @throws CRM_Core_Exception
+   */
+  public function testCreateIndividualWithJustEmail($version) {
+    $this->_apiversion = $version;
+    $params = [
+      'contact_type' => 'Individual',
+      'email' => 'garlic@example.com',
+    ];
+
+    $contact = $this->callAPISuccess('contact', 'create', $params);
```
Comment (eileenmcnaughton) [inline tests/phpunit/api/v3/ContactTest.php:1596]: This is the actual function we are testing so it's the only one that varies by api version - in general prefer apiv4

**Comment by monishdeb** [comment]
Jenkins test this please

**Comment by eileenmcnaughton** [comment]
I think this does & should fail - I don't think the actual fix is done / discussed yet

**Comment by colemanw** [comment]
Well, first off, the `'email'` field does not belong to the contact entity. So APIv4 is correct to not accept it as a parameter.
Therefore I think the expectations of this test are faulty. The expectation should be that if you create a contact with no name, and then add an email, that becomes the display name.

Hmm, I'm not sure if either version of the api will let you do that though. I'll look...

File: tests/phpunit/api/v3/ContactTest.php [hunk 2]
```diff
@@ -1573,6 +1576,38 @@ public function testCreateIndividualWithNameEmail() {
...
+   *
+   * @throws API_Exception
+   * @throws CRM_Core_Exception
+   */
+  public function testCreateIndividualWithJustEmail($version) {
+    $this->_apiversion = $version;
+    $params = [
+      'contact_type' => 'Individual',
+      'email' => 'garlic@example.com',
+    ];
```
Comment (colemanw) [inline tests/phpunit/api/v3/ContactTest.php:1594]:
```suggestion
    $params = ['contact_type' => 'Individual'];
    if ($version == 3) {
      $params['email'] = 'garlic@example.com';
    }
    else {
      $params['api.Email.create'] = ['location_type_id' => 1, 'email' => 'garlic@example.com'];
    }
```

Comment (eileenmcnaughton) [reply tests/phpunit/api/v3/ContactTest.php:1594]: Really we don't need to add test cover for v3 - the test was to demonstrate the gap/ difference

Comment (colemanw) [reply tests/phpunit/api/v3/ContactTest.php]: Ok, let's just see if this passes.

**Comment by colemanw** [comment]
Ok, test now shows that the contact is created but adding an email does not update the display name. I've confirmed that also in the APIv4 explorer.

I cannot for the life of me find where APIv3 takes care of setting the display name. Civi core appears to do it in the form layer :(

https://github.com/civicrm/civicrm-core/blob/bc77d7c0798081f999e6eca92d75fa874fd04e74/CRM/Contact/Form/Inline/Email.php#L173-L184

**Comment by eileenmcnaughton** [comment]
@colemanw I'm a bit nervous about added an extra query to every email create & update - if we can avoid it

**Comment by colemanw** [comment]
Agreed, but do you know where/how it's handled in v3?

**Comment by eileenmcnaughton** [comment]
@colemanw because 'email' is in $params the BAO handles it - the exact place is the link in the PR template I belive

**Comment by eileenmcnaughton** [comment]
Closing & returning to gitlab as this is not reviewable https://lab.civicrm.org/dev/core/-/issues/2622

**Pull request closed without merging**